File: src/Ejercicio8.cpp
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

class Biblioteca {
private:
    string nombreArchivo;

public:
    Biblioteca(string archivo) {
        nombreArchivo = archivo;
        // Asegura que el archivo exista creándolo si no está
        ofstream archivoVerificar(nombreArchivo.c_str(), ios::app);
        archivoVerificar.close();
    }

    void registrarLibro() {
        string codigo, titulo, autor, estado;
        
        cout << "\n--- Registrar Nuevo Libro ---" << endl;
        cout << "Ingrese Codigo: "; cin >> codigo;
        cout << "Ingrese Titulo (Use-guiones-sin-espacios): "; cin >> titulo;
        cout << "Ingrese Autor (Use-guiones-sin-espacios): "; cin >> autor;
        cout << "Estado (Disponible/Prestado): "; cin >> estado;

        ofstream archivo(nombreArchivo.c_str(), ios::app);
        if (archivo.is_open()) {
            archivo << codigo << " " << titulo << " " << autor << " " << estado << endl;
            archivo.close();
            cout << "-> Libro registrado exitosamente en el archivo." << endl;
        } else {
            cout << "-> Error al abrir el archivo para escribir." << endl;
        }
    }

    void mostrarPorEstado(string estadoBuscado) {
        ifstream archivo(nombreArchivo.c_str());
        string codigo, titulo, autor, estado;
        bool encontrado = false;

        cout << "\n===================================" << endl;
        cout << "       LIBROS " << estadoBuscado << "S" << endl;
        cout << "===================================" << endl;

        if (archivo.is_open()) {
            // Lee fila por fila mapeando las columnas del archivo txt
            while (archivo >> codigo >> titulo >> autor >> estado) {
                if (estado == estadoBuscado) {
                    cout << "ID: " << codigo << " | " << titulo << " - Autor: " << autor << endl;
                    encontrado = true;
                }
            }
            archivo.close();
            
            if (!encontrado) {
                cout << "No se encontraron libros con el estado: " << estadoBuscado << endl;
            }
        } else {
            cout << "Error al abrir el archivo para lectura." << endl;
        }
        cout << "===================================" << endl;
    }
};

int main() {
    Biblioteca miBiblioteca("biblioteca.txt");
    int opcion;

    do {
        cout << "\n--- MENU BIBLIOTECA VIRTUAL ---" << endl;
        cout << "1. Registrar Libro" << endl;
        cout << "2. Mostrar Libros Disponibles" << endl;
        cout << "3. Mostrar Libros Prestados" << endl;
        cout << "4. Salir" << endl;
        cout << "Seleccione una opcion: ";
        cin >> opcion;

        switch (opcion) {
            case 1:
                miBiblioteca.registrarLibro();
                break;
            case 2:
                miBiblioteca.mostrarPorEstado("Disponible");
                break;
            case 3:
                miBiblioteca.mostrarPorEstado("Prestado");
                break;
            case 4:
                cout << "Saliendo del sistema de biblioteca..." << endl;
                break;
            default:
                cout << "Opcion invalida. Intente de nuevo." << endl;
        }
    } while (opcion != 4);

    return 0;
}

File: src/Ejercicio8.java
import java.io.*;
import java.util.Scanner;

class Biblioteca {
    private String archivo = "biblioteca.txt";

    public void registrarLibro(Scanner scanner) {
        System.out.print("Código: ");
        String cod = scanner.nextLine().trim();
        System.out.print("Título: ");
        String tit = scanner.nextLine().trim();
        System.out.print("Autor: ");
        String aut = scanner.nextLine().trim();
        System.out.print("Estado (Disponible/Prestado): ");
        String est = scanner.nextLine().trim();

        if (!est.equalsIgnoreCase("Disponible") && !est.equalsIgnoreCase("Prestado")) {
            System.out.println("Estado inválido. Usa Disponible o Prestado.");
            return;
        }

        est = est.substring(0, 1).toUpperCase() + est.substring(1).toLowerCase();

        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(archivo, true));
            bw.write(cod + " " + tit + " " + aut + " " + est + "\n");
            bw.close();
            System.out.println("Libro guardado.");
        } catch (IOException e) {
            System.out.println("Error al guardar: " + e.getMessage());
        }
    }

    public void mostrarPorEstado(String estadoBuscado) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(archivo));
            String linea;
            System.out.println("\n--- Libros " + estadoBuscado + "s ---");
            while ((linea = br.readLine()) != null) {
                if (linea.contains(estadoBuscado)) {
                    System.out.println(linea);
                }
            }
            br.close();
        } catch (IOException e) {
            System.out.println("No se pudo leer el archivo (puede que esté vacío).");
        }
    }
}

public class MainBiblioteca {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Biblioteca biblio = new Biblioteca();
        int opcion = 0;

        do {
            System.out.println("\n--- Biblioteca ---");
            System.out.println("1. Registrar 2. Disponibles 3. Prestados 4. Salir");
            System.out.print("Opción: ");
            String entrada = scanner.nextLine().trim();

            try {
                opcion = Integer.parseInt(entrada);
            } catch (NumberFormatException e) {
                System.out.println("Por favor ingresa un número válido.");
                continue;
            }

            switch (opcion) {
                case 1:
                    biblio.registrarLibro(scanner);
                    break;
                case 2:
                    biblio.mostrarPorEstado("Disponible");
                    break;
                case 3:
                    biblio.mostrarPorEstado("Prestado");
                    break;
                case 4:
                    System.out.println("Saliendo...");
                    break;
                default:
                    System.out.println("Opción no válida.");
            }
        } while (opcion != 4);
        scanner.close();
    }
}
